Allow option for more verbose JSON errors

yaacovCR opened the pull request:
A la express-js/api-error-handler (https://github.com/expressjs/api-error-handler), via http-errors (https://github.com/jshttp/http-errors) -style errors objects.

ericelliott commented:
This introduces serious security risks in production apps. The example demonstrates logging the full error while returning a sanitized representation for end users.

yaacovCR commented:
Yikes. I apologize, I was planning on enabling the verbose option only for client errors, just as in api-error-handler. The sanitized version for client errors would be more verbose, however, still grabbing actual properties of the error object. How would you feel about that?

I'm new to all this, and perhaps a bit too focused on "the right way" of doing things, but what I am thinking is that client-side errors should self-document why they errored in the response. Perhaps you do this prior to the error handler, for example, by sending a more verbose json response on validation? If so, then I am not quite sure why your client-side errors make it to your error handler in general...

Just trying to wrap my head around some of this, if you have the time...

ericelliott commented:
You could use the custom serializer feature to accomplish the same. A small change would enable that:

```js
body = (o.serializer) ?
      o.serializer(mixIn({}, body, err)) :
      body;
```

No need for a verbose option, then, just pass in a custom serializer and do anything you want.

Documentation should probably warn against exposing too much info for non-user errors, and show an example of using `isClientError()` to limit error info.

Would you like to resubmit using that approach, instead?

An example in the docs and unit tests would be appreciated. =)

ericelliott commented:
Can you provide some example responses using this branch?

For example, what do you see when you hit the error route in the example?

I don't see any unit tests.

yaacovCR commented:
Gave it a go. Working for me in my project. Thank you!

ericelliott commented on error-handler.js:
```diff
@@ -98,15 +99,16 @@ var mixIn = require('mout/object/mixIn'),
 
-  send = function send(statusCode, err, res, o) {
+  send = function send(statusCode, err, res, o) {    
+    if (!err.message) err.message = statusCodes[statusCode];
```
Why are you mutating the original error object? What was wrong with the short circuit code I had?

yaacovCR replied:
Oops I documented a bit in the longer change description, basically ran into problems with mout, mixin call you suggested did not copy name from http-errors error object because that is non enumerable, so I made error the target and changed to fillin, but then ran into problem as fillin will not overwrite empty message string on error object as fillin does a null check, not a falsy check.

ericelliott replied:
Ok, but that doesn't explain why you're mutating the original error object... Would something like this work?

```js
send = function send(statusCode, err, res, o) {    
  var base = {
    name: err.name,
    status: statusCode
  };

  if (!err.message) base.message = statusCodes[statusCode];

  var body = (o.serializer) ?
    o.serializer(mixIn({}, base, err)) :
    base;

  res.status(statusCode);
  res.send(body);
}
```

yaacovCR replied:
Oh, mout!

In this version, error name is included by default, i.e. even if no serializer is defined. This is new behavior. Also, any non-enumerables or non-own properties on err are not given to serializer, except for the explicit exception of err.name. Seems kind of like a one-off.

Let me throw it back to you for a second...why is it improper to mutate the err object at this point? We are not calling next(err)... Is anything else going to inspect the err object after we handle this error? Moreover, we are adding a few properties only if they are falsy ('' for err.message and undefined for anything else), not overwriting anything. Are we mutating the error object, or possibly just augmenting it as we handle it?

ericelliott replied:
Who's idea was it to make basic properties of the error object non-enumerable in the first place? =)

> why is it improper to mutate the err object at this point?

`err` is passed as a reference. We didn't create the object, and we don't know if/how it may be used elsewhere (short of digging into the source code to find out, and the source code could change in later versions).

If we mutate it, we could break code we don't even know about. Maybe it's accessed during the `server.close()` operation? Keep in mind, this lib works in two different frameworks, so double that danger.

>  Is anything else going to inspect the err object after we handle this error?

Since we don't own the object, the answer to this question is unknowable. That's the trouble with shared mutable state...

> Moreover, we are adding a few properties only if they are falsy ('' for err.message and undefined for anything else), not overwriting anything. Are we mutating the error object, or possibly just augmenting it as we handle it?

Because we can't examine code we don't know about, we have no idea how the mutation might impact that code... Sure, I agree that deleting an expected property is more likely to cause problems than adding properties, but if there is any code checking for example, the length of `Object.keys()` (admittedly, a silly, brittle thing to do), it would break that code.

But since this lib is used in high-availability production applications with millions of monthly active users... this mutation could cause costly problems.

ericelliott replied:
How about this?

```
send = function send(statusCode, err, res, o) {    
  var base = {
    status: statusCode
  };
  var extended = {};

  if (!err.message) base.message = statusCodes[statusCode];
  if (err.name) extended.name = err.name;

  var body = (o.serializer) ?
    o.serializer(mixIn({}, err, base, extended)) :
    base;

  res.status(statusCode);
  res.send(body);
}
```

IMO, hardcoding a key whitelist is more responsible than blindly mutating the `err` object, particularly since we're just talking about what is being sent to the user. The full error can be inspected and logged elsewhere by other middleware for other purposes.

yaacovCR replied:
That works, but I like create object use better as it avoids whitelist.

On Wed, Mar 9, 2016, 11:53 PM Eric Elliott <notifications@github.com> wrote:

> In error-handler.js
> <https://github.com/ericelliott/express-error-handler/pull/12#discussion_r55633932>
> :
>
> > @@ -98,15 +99,16 @@ var mixIn = require('mout/object/mixIn'),
> >      stream.pipe(res);
> >    },
> >
> > -  send = function send(statusCode, err, res, o) {
> > +  send = function send(statusCode, err, res, o) {
> > +    if (!err.message) err.message = statusCodes[statusCode];
>
> How about this?
>
> send = function send(statusCode, err, res, o) {
>   var base = {
>     status: statusCode
>   };
>   var extended = {};
>
>   if (!err.message) base.message = statusCodes[statusCode];
>   if (err.name) extended.name = err.name;
>
>   var body = (o.serializer) ?
>     o.serializer(mixIn({}, err, base, extended)) :
>     base;
>
>   res.status(statusCode);
>   res.send(body);
> }
>
> IMO, hardcoding a key whitelist is more responsible than blindly mutating
> the err object, particularly since we're just talking about what is being
> sent to the user. The full error can be inspected and logged elsewhere by
> other middleware for other purposes.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ericelliott/express-error-handler/pull/12/files#r55633932>
> .
>

ericelliott commented on test/runtests.js:
```diff
@@ -391,3 +391,73 @@ test('JSON with serializer',
+    handler(e, testReq(), res, testNext);
+  });
+
+test('JSON with intelligent serializer that utilizes isClientError internally', 
```
`isClientError()` already has tests. This test seems to be testing... what? itself...?

yaacovCR replied:
Basically testing itself or the serializer I wrote...

ericelliott replied:
So it's not testing any actual library code that isn't already tested... So do we really need it here?

yaacovCR replied:
Nope.

yaacovCR replied:
Removed.

ericelliott commented:
Thanks a lot. This looks much better! And it was less code, and provides even more flexibility than the original PR. How do you feel about it? =)

yaacovCR commented:
Looking good to me.

yaacovCR commented:
Changed code to no longer mutate error object and deleted unnecessary test. Tests pass and works for my project but not sure if you like create object call...

yaacovCR commented:
Just touching base. Does this work for you? I am struggling a bit as mout does not seem to have a clone method for non pojos, and not truly clear to me how to use clonewith function if we don't know what kind of object will be passed.

Your comment about name being non enumerable as 'bad' idea is also well taken. Perhaps I should just not rely on its presence...

I think your solution in comment thread also works fine for me as well if you prefer that way. Just let me know and I will update accordingly.

Thanks for this library, your advice/encouragement!

ericelliott commented:
This seems fine to me, now.  Thanks again!
